lib/scholar/cluster/gmm.ex

```diff
@@ -0,0 +1,407 @@
+            run <= num_runs do
+        {{lower_bound, params}, _} =
+          while {
+                  {
+                    lower_bound = Nx.Constants.neg_infinity(),
+                    params =
+                      init_params(key, x, num_gaussians: num_gaussians, reg_covar: reg_covar)
+                  },
+                  {
+                    iter = 1,
+                    converged = Nx.tensor(false),
+                    x,
+                    key,
+                    max_iter,
+                    tol,
+                    reg_covar
+                  }
+                },
+                iter <= max_iter and not converged do
+            prev_lower_bound = lower_bound
+            {weights, means, _, precisions_cholesky} = params
+            {log_prob_norm, log_resp} = e_step(x, weights, means, precisions_cholesky)
+            params = m_step(x, log_resp, reg_covar)
+            lower_bound = log_prob_norm
+            change = lower_bound - prev_lower_bound
+            converged = Nx.abs(change) < tol
+
+            {
+              {lower_bound, params},
+              {iter + 1, converged, x, key, max_iter, tol, reg_covar}
+            }
+          end
+
+        {max_lower_bound, best_params} =
+          if max_lower_bound < lower_bound do
+            {lower_bound, params}
+          else
+            {max_lower_bound, best_params}
+          end
+
+        {
+          best_params,
+          {run + 1, max_lower_bound, x, key, max_iter, tol, reg_covar}
+        }
+      end
+
+    {weights, means, covariances, precisions_cholesky} = best_params
+
+    %__MODULE__{
+      weights: weights,
+      means: means,
+      covariances: covariances,
+      precisions_cholesky: precisions_cholesky
+    }
+  end
+
+  defnp init_params(key, x, opts) do
+    type = Nx.type(x)
+    {num_samples, _} = Nx.shape(x)
+    num_gaussians = opts[:num_gaussians]
+    reg_covar = opts[:reg_covar]
+
+    k_means_model =
+      Scholar.Cluster.KMeans.fit(x, num_clusters: num_gaussians, num_runs: 1, key: key)
+
+    labels = k_means_model.labels
+    resp = Nx.take(Nx.eye(num_gaussians, type: type), labels)
+
+    {weights, means, covariances, precisions_cholesky} =
+      estimate_gaussian_parameters(x, resp, reg_covar)
+
+    weights = weights / num_samples
+    {weights, means, covariances, precisions_cholesky}
+  end
+
+  defnp e_step(x, weights, means, precisions_cholesky) do
+    {log_prob_norm, log_resp} = estimate_log_prob_resp(x, weights, means, precisions_cholesky)
+    {Nx.mean(log_prob_norm), log_resp}
+  end
+
+  defnp m_step(x, log_resp, reg_covar) do
+    {weights, means, covariances, precisions_cholesky} =
+      estimate_gaussian_parameters(x, Nx.exp(log_resp), reg_covar)
+
+    weights = weights / Nx.sum(weights)
+    {weights, means, covariances, precisions_cholesky}
+  end
+
+  defnp estimate_gaussian_parameters(x, resp, reg_covar) do
+    nk = Nx.sum(resp, axes: [0])
+    means = Nx.dot(Nx.transpose(resp), x) / nk
```

Comment:
it might be more numerically stable to do:

```suggestion
    means = Nx.dot(Nx.transpose(resp) / nk, x / nk)
```

Reply:
Also, if `resp` can be complex, you might want Nx.LinAlg.adjoint instead

Reply:
`resp` are probabilities, so `Nx.transpose` will suffice. Should it be `Nx.sqrt(nk)` in the second expression? Is this really more numerically stable?

Reply:
If resp are probabilities then it won't make much difference, but you can just divide the other term by nk instead.

My concern was that for 2 unbounded tensors, the dot product could blow up a bit, and the division tames that to some degree.

Since one of those is bounded between 0 and 1 in each element, this isn't that much of a concern

Reply:
Indeed, dot product can explode, but here it should not be a problem. Rewriting it as
` means = Nx.dot(Nx.transpose(resp), x / nk)` might be a good idea though.